Add option to prepend each line of the output with the taske name

flyon opened the pull request:
adds support for a new config option `logTaskName`.
When set to true and used together with `logConcurrentOutput` it will prepend the taskname in a task specific color to each line of output with a default maximum of 12 chars.
When logTaskName is a number, it will be used as the maximum number of chars to be used for the taskname.

In the example below, the concurrent tasks "**web**pack:dev" and "**exe**c:dev" are prepended with their name and their own color with logTaskName=3 (max 3 chars)
![image](https://user-images.githubusercontent.com/341567/28431330-4eb208bc-6d84-11e7-9474-1e3555c02561.png)

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -4,16 +4,20 @@ var padStream = require('pad-stream');
 var async = require('async');
 var arrify = require('arrify');
 var indentString = require('indent-string');
+var colors = require('colors');
```
Use https://github.com/chalk/chalk

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +53,21 @@ module.exports = function (grunt) {
-				cp.stdout.pipe(padStream(' ', 4)).pipe(process.stdout);
-				cp.stderr.pipe(padStream(' ', 4)).pipe(process.stderr);
+				var taskLabel,numSpaces;
+				if(opts.logTaskName)
```
Follow the existing code style.

flyon commented:
hey, this is my first pull request ever .. written plenty of code, just never had to ..  I see stuff about linebreaks not being LF but CRLF, not sure what that means. Let me know if it needs fixing

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +53,21 @@ module.exports = function (grunt) {
+					numSpaces = 4;
+				}
+				cp.stdout.pipe(padStream(' ', numSpaces)).pipe(padStream(taskLabel, 1)).pipe(process.stdout);
+				cp.stderr.pipe(padStream(' ', numSpaces)).pipe(padStream(taskLabel, 1)).pipe(process.stderr);
```
Would be better to do all the padding in one `padStream` call.

flyon replied:
they use different number of repetitions tho ... so just manually generating the string of spaces first I guess?

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +53,21 @@ module.exports = function (grunt) {
+				{
+					var colorFn = colorsAvailable.length > 0 ? colors[colorsAvailable.shift()] : function(s) { return s };//use an available color or none if more tasks then colors available
+					var maxLength = typeof opts.logTaskName === 'number' ?  opts.logTaskName : Math.min(maxTaskLength,12);//task labels in output up to 12 characters
+					taskLabel = '['+colorFn(task.substr(0,maxLength))+']';
```
Use `slice` instead of `substr`.

Explanation: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/65

sindresorhus commented:
I'm ok with adding this, but needs some docs.

sindresorhus commented:
Just ignore the lint errors. I haven't updated the codebase to be compatible yet. I'll do it after merging this :)

flyon commented:
Ok, I added the requested changes and added information to the documentation.
I used String.prototype.repeat, which is Ecmascript 2015, which should be fine right?

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -4,16 +4,20 @@ var padStream = require('pad-stream');
 
 var cpCache = [];
 
+var colors = ["blue","magenta","yellow","green","red","cyan","white","gray","redBright","greenBright","yellowBright","blueBright","magentaBright","cyanBright"];
```
This needs to be inside the grunt task function since you modify it there, otherwise there will be no color left if this task is used multiple times.

sindresorhus replied:
I also don't think we should include the bright colors. They're too similar to the normal ones. Would be hard to differentiate.

flyon replied:
the way I did it now is that it uses all those colors and if all are used it uses no color. 
I thought of mapping the name to a color with color-hash, but the colors might not be distinguishable / usable. If we leave out the bright ones that would give 8 colors, and from 9 on it would be the default color. Acceptable? Or do you have a better idea?

sindresorhus replied:
I think it's fine to use color on the 8 first tasks, and then just white on the rest.

flyon replied:
cool, and I see what you mean now about it being inside the grunttask function

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +53,18 @@ module.exports = function (grunt) {
+				var padString;
+				if (opts.logTaskName) {
+					var colorFn = colors.length > 0 ? chalk[colors.shift()] : function(s) { return s };//use an available color or none if more tasks then colors available
+					var maxLength = typeof opts.logTaskName === 'number' ?  opts.logTaskName : Math.min(maxTaskLength,12);//task labels in output up to 12 characters
```
Why 12? Not sure it makes sense to crop by default at all.

flyon replied:
yeah good point, no need for any hardcoded default since its now configurable

flyon commented:
alright, all done

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -14,6 +15,8 @@ module.exports = function (grunt) {
 		});
 		var tasks = this.data.tasks || this.data;
+		var maxTaskLength = Math.max.apply(null,(tasks.map(function(task){ return task.length})));
+		var colors = ["blue","magenta","yellow","green","red","cyan","white","gray"];
```
We can't use `white` either as that's the default color in most terminals and will conflict with the non-styled tasks.

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +52,18 @@ module.exports = function (grunt) {
+				if (opts.logTaskName) {
+					var colorFn = colors.length > 0 ? chalk[colors.shift()] : function(s) { return s };//use an available color or none if more tasks then colors available
+					var maxLength = typeof opts.logTaskName === 'number' ?  opts.logTaskName : maxTaskLength;//use the longest task name as maximum, or the maximum provided by logTaskName
+					var numSpaces = (task.length > maxLength ? 0 : maxLength-task.length) + 3;//let output from all tasks be aligned
```
Instead of using `3` inline, move it to a variable with a descriptive name and use that. Makes the code more readable.

sindresorhus commented on tasks/concurrent.js:
```diff
@@ -49,8 +52,18 @@ module.exports = function (grunt) {
+					var colorFn = colors.length > 0 ? chalk[colors.shift()] : function(s) { return s };//use an available color or none if more tasks then colors available
+					var maxLength = typeof opts.logTaskName === 'number' ?  opts.logTaskName : maxTaskLength;//use the longest task name as maximum, or the maximum provided by logTaskName
+					var numSpaces = (task.length > maxLength ? 0 : maxLength-task.length) + 3;//let output from all tasks be aligned
+					padString = '['+colorFn(task.slice(0,maxLength))+']'+(' '.repeat(numSpaces));
```
The `[` and `]` should be `chalk.dim`'d

sindresorhus commented:
Can you add a test task to https://github.com/sindresorhus/grunt-concurrent/blob/master/gruntfile.js so I can just run it to manually verify that it works.

sindresorhus commented:
Ping :)

fcamblor commented:
created a new PR with your fixed suggestions : https://github.com/sindresorhus/grunt-concurrent/pull/99

This feature can be really useful to understand why some watched processes are triggered (particularly when chaining `watch`es)

sindresorhus commented:
Closing in favor of https://github.com/sindresorhus/grunt-concurrent/pull/99
